[Sai Sumanth] iP

s7manth opened the pull request:
# Dooke
> Your personal CLI assistant who is always there for you ALWAYS

```
Hello from

8888b.   dP"Yb   dP"Yb  88  dP 888888
8I  Yb  dP   Yb dP   Yb 88odP  88__
8I  dY  Yb   dP Yb   dP 88"Yb  88""
8888Y"   YbodP   YbodP  88  Yb 888888

-------------
Hello! I'm Dooke
 What can I do for you?
- - - - - - -
```

Dooke takes care of all your stuff : 
- Supports different types of tasks : todo(s), deadline(s), and event(s)
- Filters you tasks by finding matching words in them
- Marks your tasks as complete/incomplete
- Provides short descriptions/notes so that you don't forget the granularities

All you have to do is : 
1. Download it from the releases [*here*](https://github.com/s7manth/ip/releases)
2. Run `java -jar Dooke.jar`
3. Add your tasks
4. Let it manage your tasks for you 😉 

- [x] Do it in **Java**!
- [ ] For the people who love **Java**!
- [x] If you like ~Python~, you are not my friend!

``` java 
// If you know java, you should know this 

public class Main {
    public static void main(String[] args) {
        System.out.println("Hello World!");
    }
}    

```

Commit:
Add Gradle support

s7manth committed:
add .class to ignore java bytecode files

s7manth committed:
add basic input/output manager, outputs verbatim input and exit on bye

s7manth committed:
add task and tasklist implementation to maintain history

s7manth committed:
add input handler, type and validator to process input commands

s7manth committed:
add custom string output formatter for multiple string appends at once

s7manth committed:
add support for mark, unmark commands to indicate completion status, and formatted output

s7manth committed:
add multiple types of tasks

s7manth committed:
add support for todo, event, and deadline tasks

s7manth committed:
edit name of the cli chatbot

s7manth committed:
add test case for testing ui and edited the shell script to support the renamed file

s7manth committed:
remove unnecessary whitespaces to pass ui tests

s7manth committed:
rename input validator as identifier, and the package to tasks instead of task

s7manth committed:
add different types of exceptions

s7manth committed:
complete exception handling for todo, mark, unmark, and unknown commands

s7manth committed:
add deadline and event related exceptions

s7manth committed:
handle invalid deadline and event commands using exceptions

s7manth committed:
add simple test to check exception behavior

s7manth committed:
add delete command to remove tasks from the list

s7manth committed:
add exception handling for delete command and variants

s7manth committed:
add exception to handle corrupt files

s7manth committed:
add support for persistent storage of tasks in a file and handle related exceptions

s7manth committed:
add localdate and an optional localtime support for deadline, event tasks

s7manth committed:
Merge branch 'branch-Level-8'

s7manth committed:
move tasklist to tasks package

s7manth committed:
add basic junit tests

s7manth committed:
add java doc comments to core and utilities packages excluding tasks and exceptions

s7manth committed:
fix switch block indentation to abide by the coding standard

s7manth committed:
add support to search through the task description to filter tasks and handle related exceptions

s7manth committed:
Merge branch 'branch-A-CodingStandard'

s7manth committed:
Merge branch 'branch-Level-9'

s7manth committed:
complete java doc comments for all the classes and rename DukeException to DookeException

s7manth committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-Add-Gradle

s7manth committed:
add gradle tool to automate junit testing, cleaning unused binaries

s7manth committed:
add config files for running checkstyle

s7manth committed:
fix checkstyle errors and warnings

s7manth committed:
set up basic javafx stage and scene

brelkh commented on src/main/java/core/tasks/Deadline.java:
```diff
@@ -0,0 +1,98 @@
+     * @param description The description of the deadline.
+     * @param by The time by which the deadline is supposed to be achieved.
+     */
+    private Deadline(String description, String by) {
+        super(description);
+        String[] dateAndTime = by.split(" ");
+        this.byDate = LocalDate.parse(dateAndTime[0]);
+        if (dateAndTime.length > 1) {
+            this.byTime = LocalTime.parse(dateAndTime[1]);
+        }
+        this.by = by;
+    }
```
Would a variable name such as "endDate" or "endTime" be more descriptive to readers than "by"?

brelkh commented on src/main/java/core/tasks/Deadline.java:
```diff
@@ -0,0 +1,98 @@
+    /**
+     * Factory method to obtain an instance of the Deadline class.
+     *
+     * @param description The description of the Deadline.
+     * @param by The time by which the deadline must be achieved.
+     * @return An instance of the Deadline class with the specified description and by time.
+     * @throws NoDeadlineMentionedException Throws an exception if no deadline is mentioned.
+     * @throws NoDescriptionGivenException Throws an exception if the description is empty or blank.
+     * @throws DateTimeParseException Throws an exception if the deadline entered is in an incorrect format.
```
Just wanted to commend your consistency with comments and javadocs! They are clear, easy to understand, and pretty much every method has an appropriate javadoc. 👍

brelkh commented on src/main/java/core/tasks/Event.java:
```diff
@@ -0,0 +1,98 @@
+public class Event extends Task {
+    protected LocalDate atDate;
+    protected LocalTime atTime;
+    protected String at;
+
+    /**
+     * Constructor for the Event class.
+     *
+     * @param description The description of the event.
+     * @param at The locale of the event.
+     */
```
Similar to "by" for the deadline class, perhaps using the word "locale" like you used in the javadoc comment would be more clearly understood than "at"?

brelkh commented on src/main/java/core/tasks/TaskList.java:
```diff
@@ -0,0 +1,231 @@
+ * @author s7manth
+ * @version 0.1
+ */
+public class TaskList {
+    private ArrayList<Task> taskList;
+
```
As suggested in the java coding standard, you could consider using the plural form of the object to represent the collection of objects. ie. `tasks` instead of `taskList` unless there are various collection types for the task object like `taskSet` and `taskList`.

brelkh reviewed (commented):
Overall, your code LGTM with regards to java coding standards. I tried to look for more basic violations like wrong whitespacing or unclear package statements but could not find any. I also appreciate your commitment to ensuring clear javadoc comments in every method. :)

MelvinCZJ commented on src/main/java/core/InputHandler.java:
```diff
@@ -90,6 +90,8 @@ public String handleInput(String inputData) {
         case UNKNOWN :
             handleUnknown(outputFormatter);
```
I like the names you give your methods! 👍

MelvinCZJ commented on src/main/java/core/InputHandler.java:
```diff
@@ -254,9 +264,10 @@ private void handleDelete(String inputData, OutputFormatter outputFormatter) {
+                    "Now you have ", taskList.getLength(), " task(s) in the list.");
 
-        } catch ( NumberFormatException e) {
+        } catch (NumberFormatException e) {
```
Before I read anymore of your code, I realized how much I am lacking in terms of exception handling. I essentially lumped all exceptions under a DukeException class, except that I modified the exception message to describe the specific reason the exception was thrown. In hindsight, I probably should have created individual Exception subclasses to handle each type of exception.

MelvinCZJ commented on src/main/java/core/IoManager.java:
```diff
@@ -1,71 +1,74 @@
  * @version 0.1
  */
-public class IOManager {
+public class IoManager {
```
This seems like an interesting method of breaking down the components of Dooke!

MelvinCZJ commented on src/main/java/core/tasks/Deadline.java:
```diff
@@ -85,7 +86,8 @@ public static Deadline getInstance(String description, String by) throws NoDeadl
-        outputFormatter.appendAll("[D]", super.toString(), " (by: ", byDate.format(DateTimeFormatter.ofPattern("MMM d yyyy")));
+        outputFormatter.appendAll("[D]", super.toString(), " (by: ",
+                byDate.format(DateTimeFormatter.ofPattern("MMM d yyyy")));
         if (byTime != null) {
```
Perhaps there could be another way of handling this check other than using a null check? 🤔

MelvinCZJ commented on src/main/java/Dooke.java:
```diff
@@ -7,16 +7,20 @@
         System.out.println("Hello from\n" + logo);
 
-        IOManager ioManager = IOManager.getInstance();
+        IoManager ioManager = IoManager.getInstance();
```
Not sure where would be a good place to mention this but, while I do think that the naming convention you follow is great, but I'm think it could be hard to understand for some. At the same time, I do also understand that the code will most likely be viewed by people who can understand it in the first place. (I'm assuming people who use github are people who are able to read and understand codes to a considerable extent.) 💭

MelvinCZJ reviewed (commented):
Overall, code is pretty clean, and compliant to coding standard. (I tried really hard to spot any mistakes but I have also been awake for almost 24-hours straight now. Not trying to make any excuses here!😢) I would also like to mention, regarding the comment on the naming convention, I'm not sure if anyone else shares the same opinion as me, but I would like to believe that the fact that I find your naming conventions a little tough to understand might be also be due to the fact that you might be an intermediate to advanced level programmer? 🥲

s7manth committed:
add minimal dooke gui

s7manth committed:
remove unnecessary images

s7manth committed:
add missing java docs and fix checkstyle errors

s7manth committed:
add author and version details for all classes

s7manth committed:
update readme

s7manth committed:
fix code quality issues, mainly pertaining to java docs

s7manth committed:
use assert feature to document assumptions

Not using assertions at various code checkpoints like the creation of
the file and empty input commands to ensure that assumptions hold.

Adding assertions allows to properly manage the assumptions and provide
a feedback system if the code stops working.

Important code fragments that depend on external system are being
asserted of the presence of a particular assumption in place before
performing any operations.

Assert checks whether the assumptions made by the developer are correct
or not.

s7manth committed:
Merge pull request #2 from s7manth/branch-Add-Assertions

use assert feature to document assumptions

s7manth committed:
perform critical code checks and improve the code quality

The common behaviors across deadline and event input handlers cause code duplication.

Extracting the common behaviors into a method allows us to abstract out
the granularities of checking the input command.

Let's pull up behaviors common to the method which checks and
initializes important parameters for the tasks.

Making a method is preferrable here because there are only two types of
tasks that make use of this parameter check to avoid invalid commands.

s7manth committed:
Merge pull request #3 from s7manth/branch-A-CodeQuality

perform critical code checks and improve the code quality

s7manth committed:
add continuous integration to the repository

Proper github workflow is not present for the repository.

Adding gradle.yml file for github to perform CI checks and run the
workflow every time certain project events are triggered.

CI checks provides features like code review and code quality checks
making identification of errors easy.

A highly recommended DevOps tool for ensuring project's proper life
cycle.

s7manth committed:
fix checkstyle errors in test module

s7manth committed:
edit platform to macos

s7manth committed:
Merge pull request #4 from s7manth/branch-Add-CI

add continuous integration to the repository

s7manth committed:
add help command support

Chatbot does not contain a help command for the user to understand how
to input commands.

Help command provides details regarding the input command format for the
user such that they do not run into trouble while using the chatbot.

Adding support for the help command by specifying the various commands
supported by the chatbot and their respective acceptable formats.

s7manth committed:
polish the gui

s7manth committed:
add missing java docs

s7manth committed:
add user guide

s7manth committed:
Set theme jekyll-theme-hacker

s7manth committed:
change version number

s7manth committed:
fix the path reference to main class and runtest arguments

s7manth committed:
update readme to include more details

s7manth committed:
Create CNAME

s7manth committed:
Delete CNAME

s7manth committed:
Set theme jekyll-theme-hacker

s7manth committed:
scale the image to width 250
